src/wallet/crypter.cpp

```diff
@@ -185,24 +185,22 @@ static bool DecryptSaplingSpendingKey(const CKeyingMaterial& vMasterKey,
     return sk.expsk.full_viewing_key() == extfvk.fvk;
 }
 
+// cs_KeyStore lock must be held by caller
 bool CCryptoKeyStore::SetCrypted()
 {
-    LOCK2(cs_KeyStore, cs_SpendingKeyStore);
     if (fUseCrypto)
         return true;
     if (!(mapKeys.empty() && mapSproutSpendingKeys.empty() && mapSaplingSpendingKeys.empty()))
         return false;
-    fUseCrypto = true;
-    return true;
+    return fUseCrypto = true;
 }
 
 bool CCryptoKeyStore::Lock()
 {
-    if (!SetCrypted())
-        return false;
-
     {
         LOCK(cs_KeyStore);
+        if (!SetCrypted())
+            return false;
         vMasterKey.clear();
     }
 
```

Comment:
For performance reasons, calls to `NotifyStatusChanged()` should be made outside the lock, as is done here (and throughout this file, well, there's only one other place).

Reply:
There might also be locking order reasons to call it outside the lock. (I'm not sure, but it looks like a cleaner contract to let the notification handler decide what to lock and in what order.)